network/noise/src/noise_fuzzing.rs

```diff
@@ -0,0 +1,293 @@
+        context: &mut Context,
+        buf: &[u8],
+    ) -> Poll<futuresResult<usize>> {
+        // write to us first
+        self.written.extend_from_slice(buf);
+        //
+        Pin::new(&mut self.inner).poll_write(context, buf)
+    }
+
+    fn poll_flush(mut self: Pin<&mut Self>, context: &mut Context) -> Poll<futuresResult<()>> {
+        Pin::new(&mut self.inner).poll_flush(context)
+    }
+
+    /// Attempt to close the channel. Cannot Fail.
+    fn poll_close(mut self: Pin<&mut Self>, context: &mut Context) -> Poll<futuresResult<()>> {
+        Pin::new(&mut self.inner).poll_close(context)
+    }
+}
+impl AsyncRead for ExposingSocket {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        context: &mut Context,
+        buf: &mut [u8],
+    ) -> Poll<futuresResult<usize>> {
+        Pin::new(&mut self.inner).poll_read(context, buf)
+    }
+}
+
+// take_socket is implemented on parent structure to be able to retrieve our ExposingSocket
+impl NoiseSocket<ExposingSocket> {
+    fn take_socket(self) -> ExposingSocket {
+        self.socket
+    }
+}
+
+//
+// Actual code to generate corpus
+//
+
+// let's cache the deterministic keypair
+use once_cell::sync::Lazy;
+pub static NOISE_KEYPAIR: Lazy<snow::Keypair> = Lazy::new(|| {
+    let parameters: NoiseParams = NOISE_PARAMETER.parse().unwrap();
+    Builder::with_resolver(parameters.clone(), Box::new(TestResolver {}))
+        .generate_keypair()
+        .unwrap()
+});
+
+fn generate_first_two_messages() -> (Vec<u8>, Vec<u8>) {
+    // build
+    let parameters: NoiseParams = NOISE_PARAMETER.parse().unwrap();
+    let initiator = snow::Builder::new(parameters.clone())
+        // why not use the same as the responder :D
+        .local_private_key(&NOISE_KEYPAIR.private)
+        .remote_public_key(&NOISE_KEYPAIR.public)
+        .build_initiator()
+        .unwrap();
+    let responder = snow::Builder::new(parameters)
+        .local_private_key(&NOISE_KEYPAIR.private)
+        .build_responder()
+        .unwrap();
+
+    // create exposing socket
+    let (dialer_socket, listener_socket) = ExposingSocket::new_pair();
+    let (dialer, listener) = (
+        NoiseSocket::new(dialer_socket, initiator),
+        NoiseSocket::new(listener_socket, responder),
+    );
+
+    let (dialer, listener) = (Handshake(dialer), Handshake(listener));
+
+    let (dialer_result, listener_result) =
+        block_on(join(dialer.handshake_1rt(), listener.handshake_1rt()));
+
+    // take result
+    let init_msg = dialer_result.unwrap().take_socket().written;
+    let resp_msg = listener_result.unwrap().take_socket().written;
+
+    (init_msg, resp_msg)
+}
+
+pub fn generate_corpus(gen: &mut libra_proptest_helpers::ValueGenerator) -> Vec<u8> {
+    let (init_msg, resp_msg) = generate_first_two_messages();
+    // choose a random one
+    let strategy = proptest::arbitrary::any::<bool>();
+    if gen.generate(strategy) {
+        init_msg
+    } else {
+        resp_msg
+    }
+}
```

Comment:
will this ever generate more than two different messages? I guess the ephemeral keys will be different?

Reply:
we don't really care about generating more than two messages actually, but this is the API we have in the fuzzer framework :D